### PR TITLE
SLM-193: Save the source IP for all barcode event types

### PR DESCRIPTION
**Pull request by nathanrussell-moj-digital**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeResource.kt [hunk 1]
```diff
@@ -60,10 +61,11 @@ class BarcodeResource(private val barcodeService: BarcodeService, private val us
   )
   fun createBarcode(
     @Parameter(hidden = true) @AuthenticationPrincipal userDetails: UserDetails,
+    @Parameter(hidden = true) @SourceIpAddress sourceIp: String,
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeResource.kt:64]: I particularly like the implementation of the `@SourceIpAddress` annotation (see SourceIpAddressMethodArgumentResolver below) 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/SourceIpAddressMethodArgumentResolver.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * The web request's source IP is resolved by iterating through the well known headers that could contain the source IP.
+ * The first of these to contain a value will be returned, else [HttpServletRequest.getRequestIp] is returned.
+ * This strategy allows for proxies/load balancers etc where the [HttpServletRequest.getRequestIp] may be the IP of the
+ * load balancer rather than the originally requesting IP.
+ * The sequence of the [IP_HEADERS] list is therefore significant.
+ *
+ * If the native request is not a [HttpServletRequest] an empty string is returned.
+ */
+class SourceIpAddressMethodArgumentResolver : HandlerMethodArgumentResolver {
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/SourceIpAddressMethodArgumentResolver.kt:28]: Spring and its interfaces is ace. That is all.
 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeResource.kt (hunk 1 above)
Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeResource.kt]: Removed class as implementation approach has changed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/SourceIpAddressMethodArgumentResolver.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      "WL-Proxy-Client-IP",
+      "HTTP_X_FORWARDED_FOR",
+      "HTTP_X_FORWARDED",
+      "HTTP_X_CLUSTER_CLIENT_IP",
+      "HTTP_CLIENT_IP",
+      "HTTP_FORWARDED_FOR",
+      "HTTP_FORWARDED",
+      "HTTP_VIA",
+      "REMOTE_ADDR"
+    )
```
[the remark this answers is not in the document]
Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/config/SourceIpAddressMethodArgumentResolver.kt]: Removed class as implementation approach has changed

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**